src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -938,9 +938,17 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         //           (arity > 0) expected type is a SAM that is not annotated with `@FunctionalInterface`
         //   - 3.0: (arity == 0) expected type is a SAM that is not annotated with `@FunctionalInterface`
         def checkCanEtaExpand(): Boolean = {
-          def expectingSamOfArity = {
+          // under -Xsource:3, be more strict about checking the param and return types to avoid
+          // SAM eta-expansion getting incorrectly prioritized over auto application (scala/bug#12006)
+          def expectingSamOfArity(typeCheck: Boolean) = {
             val sam = samOf(ptUnderlying)
-            sam.exists && sam.info.params.lengthCompare(arity) == 0
+            sam.exists &&
+            (if (typeCheck)
+              sam.info match {
+                case mt2: MethodType => isApplicableMethod(mt, mt2)
```

Comment:
this should only be checked if arity is 0, no? so it would be enough to check return types.

Reply:
Yes. I went back and forth on this PR. At some point I was trying to do the check with or without -Xsource:3. I can bring the code back in here to clarify that it's only doing return type check.